### ISSUE TITLE
Question on Winograd result interpretation

### ISSUE THREAD

**Issue by flipz357** [jgroschwitz/GrAPES#4]
[sorry to bother you more, this time it's only a question, due to my excitement about the paper:-) ]

As far as I know, the WSC data set is balanced, so I'd expect the edge recall to be 50% for the random baseline, since that's the probability of a coin flip. Maybe, considering the prerequisites, and other parsing hickups, it should be lower, but 2% is strikingly low. If you find the specific edge and process it such that it holds for the other WSC option, it could be 98% accuracy then? So maybe the parser is very good for WSC, just it always does the opposite of what's right?

**Comment by jgroschwitz** [comment]
No worries! Very happy to see that you're interested in our work, and to answer your questions.

This is a good one. The thing is that there are not just two cases here (attach to correct node or to the distractor), but a third one: that there is no edge corresponding to the coreference at all. The AM parser actually falls into the latter category: it simply has no mechanism to model coreference (and all* reentrant edges due to coreference are removed in the training data). I don't remember if I looked at the AM parser predictions on Winograd specifically, but I would think that most of the 98% error cases are not attachments to the distractor, but not predicting an edge at all (or inserting a pronoun node and attaching the edge there).

AMRBart gets actually about half of the edges correct where the prerequisites are satisfied, I wonder if that means that it actually mostly does that 50% chance accuracy, or if there are other errors in there and the resolution itself is >50%. Not sure about the low performance of Cai&Lam -- I'm curious now, I might have a look if I find the time :)

(*) Or more precisely, the edges that our heuristic thinks are due to coreference.

**Comment by megodoonch** [comment]
Thanks for your thoughts! I like the idea of these comparisons, though they'd still have a lot of uncertainty. For instance, with the longer sentence maybe the parser just gets confused about something else entirely, and so Smatch changes, but not in the relavant way. This is why we have these specific metrics.

The AM parser would still not get these, though -- there's also no merging of like-labelled nodes. (And really, there shouldn't be. You wouldn't want to merge in "Mary's suitcase is larger than Sue's suitcase", right?) The AM parser is correctly "punished" for missing these kinds of coreference. Just because it's not designed for it doesn't mean it's not failing at the task, since it's supposed to be a general AMR parser. 

There is, by the way, a way to handle coreference like this in the AM Algebra, up to a finite number of additional COREF-n sources/slot names. It's just not in the parser, and I think it would introduce a lot of messiness.

**Comment by flipz357** [comment]
> with the longer sentence maybe the parser just gets confused about something else entirely, and so Smatch changes, but not in the relavant way.

Yes, you're right, the Smatch might change not in a relevant way, but the idea would be that these changes are somewhat controlled and normalized (because of the close pairs), and so the overall score on WSC might become more interpretable (the interpretability issue of the WSC score was also mentioned in your paper I think).

Though all in all, your way of measuring the WSC performance may be the best possible, just my idea above could be something to think about for checking against. It also wouldn't need a gold reference.

PS: Thanks for the info on the AM parser.

**Comment by flipz357** [comment]
True it wouldn't really be a GrAPES-style evaluation, but something different.

Wondering if there's a way to make sure that A implies not-D, but I guess there isn't.

Anyways, yeah, what you say makes sense: additionally reporting P(D | C) would be great as it provides an more interpretable score. On top of that, also reporting P(A) would then give us "safety value" and tells how "trusty" P(D|C) is. Maybe as a last thing to consider, for P(D | C) it could make sense to include B in C "50-50" correct/false, since "no-re-entrancy" could be viewed as equivalent to random guess re winograd problem.

**Comment by jgroschwitz** [comment]
Good points! One note: the way we measure things, A actually implies not-D, because if the prerequisites are not satisfied, our reentrancy metric always returns a failure.

Simplified speaking[1], for D we measure "does an edge between two nodes that are labelled x and y exist". But if no such nodes exist in the first place (which is exactly a failure on the prerequisites, i.e. case A), then that metric always returns false, no matter what edges exist in the graph. (We do it this way, because checking whether the correct reentrancy exists, when there are node errors like that, becomes very tricky and blurry. So the prerequisites essentially cover for a shortcoming of our proper metric, where it can fail for reasons that are not related to reentrancies).

The idea is that, similar to the idea with P(D|C) here, one can then look at P(D) for absolute performance, and P(D| not A) for  error rates specific to the reentrancies.

[1] Actually for reentrancies it is slightly more complex, to take the reentrancy-aspect into account more concretely (we essentially check that the other incoming edge at the reentrant node is also there; I think the appendix of the paper explains it in more detail).

**Comment by flipz357** [comment]
Interesting, thanks!

But doesn't

> So the prerequisites essentially cover for a shortcoming of our proper metric

a bit contradict with

> A actually implies not-D

Maybe we've had slightly different conception, I guess you meant "it implies" from your metric point of view, I meant more like that D would be about "the parser solve the WSC problem" (unrelated to a particular way of measuring). As I see it, the prerequisuite can fail for some reasons, but when a human would look at it afterwards, they would grade it correctly, so what I understand is A means "probably the parser didn't solve the WSC problem but we can't say for sure".

Hope I didn't put it too confusing...

**Comment by jgroschwitz** [comment]
I see! Yes, you're right. So giving the above categories a * to denote "as measured by our metric*, and without a star the "true" values, say by human judgement, then indeed A* implies not-D*, but A does not imply not-D (sorry, reaching dangerous numbers of negations within one sentence there...).

I would more say A means something like "the parser didn't get a chance to solve the WSC problem, because it already failed on something else". Might be two sides of the same coin though!

**Comment by flipz357** [comment]
> Might be two sides of the same coin

Could be, but not entirely sure yet, because sometimes there may be mutliple similar okaish ways to model something in AMR, and projected concepts can be abstract. For instance a parser may parse "boss" into (b / boss) or something like (p / person :arg2-of (have-org-role-91 (b / boss))). 

Depending on the choice, a coref might either involve "boss" or "person", and both could be right, or wrong resolutions of a WSC. But the prerequisite would (if I understand it correctly) trigger only at one option. 

That's why I suggest to interpret P(A) as a "trust measure" for the interpretable main score: (#D + (#B / 2)) / (#C + #B/2), leaving out A and showing it only as a "trust-measure" for the main score. The main score is then in a way interpretable that it would approximate the probability of a parser that it solve WSC problems correctly.

**Comment by jgroschwitz** [comment]
That makes sense! I'll include output along the lines of this discussion in the next round of updates. Thanks for your feedback :)
